engine/include/Engine/Intermediate/DataChunk.inl

```diff
@@ -0,0 +1,52 @@
+#include "Engine/Intermediate/DataChunk.hpp"
+
+template <typename TStoredComponent, int TSize>
+constexpr inline DataChunk<TStoredComponent, TSize>::DataChunk() noexcept
+{
+    m_components.reserve(TSize / sizeof(TStoredComponent));
+}
+
+template <typename TStoredComponent, int TSize>
+template <typename... Args>
+TStoredComponent& DataChunk<TStoredComponent, TSize>::addComponent(Args&&... args) noexcept
+{
+    std::unique_lock lock(m_mutex);
+
+    return m_components.emplace_back(std::forward<Args>(args)...);
+}
+
+template <typename TStoredComponent, int TSize>
+void DataChunk<TStoredComponent, TSize>::destroyComponent(const TStoredComponent* componentToDestroy) noexcept
+{
+    std::unique_lock lock(m_mutex);
+
+    for (size_t i = 0; i < m_components.size(); ++i)
+    {
+        if (unlikely(&m_components[i] == componentToDestroy))
+        {
+            std::swap(m_components[i], m_components.back());
+
+            m_components[i].getGameObject().updateComponentLink<TStoredComponent>(&m_components.back(),
+                                                                                  &m_components[i]);
+            m_components.pop_back();
+        }
+    }
+}
+
+template <typename TStoredComponent, int TSize>
+DataChunk<TStoredComponent, TSize>* DataChunk<TStoredComponent, TSize>::getInstance() noexcept
+{
+    // double same if to avoid to lock mutex
+    if (unlikely(m_pInstance == nullptr))
+    {
+        std::unique_lock lock(m_mutex);
+        if (unlikely(m_pInstance == nullptr))
+        {
+            m_pInstance = new DataChunk<TStoredComponent, TSize>();
+        }
+        return m_pInstance;
+    }
+
+    std::shared_lock lock(m_mutex);
+    return m_pInstance;
+}
```

Comment:
We could lazy-load the chunk on the stack, by moving its instance into this method (which, by the way, I suggest we rename `get()`):
```suggestion
template <typename TStoredComponent, int TSize>
DataChunk<TStoredComponent, TSize>& DataChunk<TStoredComponent, TSize>::getInstance() noexcept
{
    // IMPORTANT: if this approach is adopted, lock m_mutex in DataChunk's constructor
    // IMPORTANT: if this approach is adopted, remove m_pInstance from data members
    static DataChunk<TStoredComponent, TSize> chunk;
    return chunk;
}
```